pkg/bitwarden/cli.go

```diff
@@ -0,0 +1,161 @@
+	sd  *sessionData
+}
+
+func newBitwardenCliClient() Client {
+	return bitwardenCliClient{
+		run: defaultCliRun,
+		sd:  &sessionData{},
+	}
+}
+
+type bwLoginResponse struct {
+	Success bool `json:"success"`
+	Data    struct {
+		Raw string `json:"raw"`
+	} `json:"data"`
+}
+
+var (
+	defaultCliRun = run
+)
+
+type sessionData struct {
+	sync.Mutex
+	session    string
+	savedItems []Item
+}
+
+func (c bitwardenCliClient) Login(username, password string) error {
+	c.sd.Lock()
+	defer c.sd.Unlock()
+	output, err := c.run("bw", "login", username, password, "--response")
+	if err != nil {
+		return err
+	}
+	r := bwLoginResponse{}
+	if err := json.Unmarshal(output, &r); err != nil {
+		//TODO need to censor if we plan this on a job
+		return err
+	}
+	if r.Success {
+		raw := r.Data.Raw
+		if raw != "" {
+			c.sd.session = raw
+			return nil
+		}
+		// should never happen
+		return fmt.Errorf("bw-login succeeded with empty '.data.raw'")
+	}
+	// should never happen
+	return fmt.Errorf("bw-login failed without run's error")
+}
+
+func (c bitwardenCliClient) listItems() ([]Item, error) {
+	var items []Item
+	if c.sd.session == "" {
+		return items, fmt.Errorf("not yet login to Bitwarden")
+	}
+	if len(c.sd.savedItems) > 0 {
+		return c.sd.savedItems, nil
+	}
+	out, err := c.run("bw", "--session", c.sd.session, "list", "items")
+	if err != nil {
+		return items, err
+	}
+	err = json.Unmarshal(out, &items)
+	if err != nil {
+		return items, err
+	}
+	return items, nil
+}
+
+func (c bitwardenCliClient) GetField(itemName, filedName string) (string, error) {
+	if c.sd.session == "" {
+		return "", fmt.Errorf("not yet login to Bitwarden")
+	}
+	if len(c.sd.savedItems) == 0 {
+		if _, err := c.listItems(); err != nil {
+			return "", err
+		}
+	}
+	for _, i := range c.sd.savedItems {
+		if i.Name == itemName {
+			for _, field := range i.Fields {
+				if field.Name == filedName {
+					return field.Value, nil
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("not found field %q in item %q", filedName, itemName)
+}
+
+func (c bitwardenCliClient) GetAttachment(itemName, attachmentName string) (string, error) {
+	if c.sd.session == "" {
+		return "", fmt.Errorf("not yet login to Bitwarden")
+	}
+	if len(c.sd.savedItems) == 0 {
+		if _, err := c.listItems(); err != nil {
+			return "", err
+		}
+	}
+	for _, i := range c.sd.savedItems {
+		if i.Name == itemName {
+			for _, attachment := range i.Attachments {
+				if attachment.FileName == attachmentName {
+					//bw --session "${BW_SESSION}" get attachment "${attachment_id}" --itemid "${item_id}" --raw
+					out, err := c.run("bw", "--session", c.sd.session, "get", "attachment", attachment.ID, "--itemid", itemName, "--raw")
+					return string(out), err
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("not found attachment %q in item %q", attachmentName, itemName)
+}
+
+func (c bitwardenCliClient) Logout() error {
+	_, err := c.run("bw", "logout", "--response")
+	if err != nil {
```

Comment:
nit: you don't use `err` outside of this context so 


```go
if _, err := c.run(/*...*/); err != nil {

}
```

Reply:
Done.